featurebyte/query_graph/node/function.py

```diff
@@ -0,0 +1,227 @@
+        return self.parameters.output_dtype
+
+    def _derive_node_operation_info(
+        self,
+        inputs: List[OperationStructure],
+        branch_state: OperationStructureBranchState,
+        global_state: OperationStructureInfo,
+    ) -> OperationStructure:
+        input_category = inputs[0].output_category
+        row_index_lineage = inputs[0].row_index_lineage
+        for input_operation_structure in inputs:
+            # check input category type is homogeneous
+            if input_operation_structure.output_category != input_category:
+                raise ValueError("Input category type is not homogeneous")
+            if input_operation_structure.output_type != NodeOutputType.SERIES:
+                raise ValueError("Input type is not series")
+            if input_operation_structure.row_index_lineage != row_index_lineage:
+                raise ValueError("Input row index is not matched")
+
+        # prepare node parameters
+        columns = []
+        aggregations = []
+        for input_operation_structure in inputs:
+            columns.extend(input_operation_structure.columns)
+            aggregations.extend(input_operation_structure.aggregations)
+
+        node_kwargs: Dict[str, Any] = {"columns": []}
+        if input_category == NodeOutputCategory.VIEW:
+            node_kwargs["columns"] = [
+                DerivedDataColumn.create(
+                    name=None,
+                    columns=columns,
+                    transform=self.parameters.function_name,
+                    node_name=self.name,
+                    dtype=self.parameters.output_dtype,
+                )
+            ]
+        else:
+            node_kwargs["columns"] = columns
+            node_kwargs["aggregations"] = [
+                PostAggregationColumn.create(
+                    name=None,
+                    columns=aggregations,
+                    transform=self.parameters.function_name,
+                    node_name=self.name,
+                    dtype=self.parameters.output_dtype,
+                )
+            ]
+
+        return OperationStructure(
+            **node_kwargs,
+            output_type=NodeOutputType.SERIES,
+            output_category=input_category,
+            row_index_lineage=row_index_lineage,
+        )
+
+    def _derive_sdk_code(
+        self,
+        node_inputs: List[VarNameExpressionInfo],
+        var_name_generator: VariableNameGenerator,
+        operation_structure: OperationStructure,
+        config: CodeGenerationConfig,
+        context: CodeGenerationContext,
+    ) -> Tuple[List[StatementT], VarNameExpressionInfo]:
+        function_parameters: List[Any] = []
+        node_input_index = 0
+        for func_param in self.parameters.function_parameters:
+            func_param_val, node_input_index = func_param.get_sdk_function_argument(
+                node_inputs, node_input_index
+            )
```

Comment:
I'm not sure if the `get_sdk_function_argument()` abstraction is necessary since `ValueFunctionParameterInput` basically ignores `node_inputs` and `node_input_index`.

Wondering if something simpler like this is easier to understand:

```python
if func_param.input_form == "value":
    func_param_val = ValueStr.create(func_param.value)
else:
    func_param_val = node_inputs[node_input_index]
    node_input_index += 1
```

Reply:
This is what i did initially and it is suggested by @jevonyeoh to abstract the logic here. @jevonyeoh

Reply:
i think it's easier to extend in the current form if we ever need to add more types of inputs since we can just add another implementation, and not have to worry about updating all the if-statements.